### PR TITLE
Add someip service discovery module

### PR DESCRIPTION
**Pull request by marcelbuesing**
Adds most of missing parts of some ip sd. I think documentation and tests still needs improvement.
I am not really sure about everything requiring vectors/dynamic memory allocations. I guess an alternative could be something like [heapless::Vec](https://docs.rs/heapless/0.7.7/heapless/struct.Vec.html).

### CODE REVIEW

**Comment by JulianSchmid** [comment]
Hi @marcelbuesing,

First: Thank you for the PR. SD support is great and also thank you for all the tests you wrote. I am currently sick, but will try have a deeper look at your PR next week when I am (hopefully) feeling better.

Concerning allocations: I originally tried to avoid them at all cost, but I understand that this gets almost impossible if you enter SOMEIP SD territory. The alternative to allocations is to reserve the maximum amount of memory that could be used. In SOMEIP SD's case this is the maximum size of an UDP packet which is ~65'535 bytes (minus a few bytes for the SOMEIP header). Depending on the system you are running on or the use case reserving the maximum amount of memory might be the right or the wrong choice. So I think this is a decision that the user of the library will have to make and the library should offer different options. Additionally the maximum size would get even bigger if someone chooses to use IPV6 Jumbograms, which would blow up the maximum size to ~4GiB, which would make allocating the maximum amount quiet impossible or at least costly. While it is unlikely that someone will combine SOMEIP with IPv6 Jumbograms, I learned that you should never assume anything and that at some point someone will do it.

TLDR: I am no longer against allocations, but I was thinking of flagging all parts of the library that need allocation or std types with the std feature flag (similar to https://github.com/BurntSushi/byteorder/blob/master/Cargo.toml#L25 ). That way users of the library can make a choice if they want to use those parts or explicitly exclude them. 

So the plan would be that I review & merge your PR in the following weeks, maybe put the std feature flag on top of that and then do a release. Unless the std feature flagging takes me too long, then I will do a release just with your changes and the std feature on top later.

Greets
Julian

**Comment by marcelbuesing** [comment]
> Hi @marcelbuesing,
> 
> First: Thank you for the PR. SD support is great and also thank you for all the tests you wrote. I am currently sick, but will try have a deeper look at your PR next week when I am (hopefully) feeling better.
> 
> Concerning allocations: I originally tried to avoid them at all cost, but I understand that this gets almost impossible if you enter SOMEIP SD territory. The alternative to allocations is to reserve the maximum amount of memory that could be used. In SOMEIP SD's case this is the maximum size of an UDP packet which is ~65'535 bytes (minus a few bytes for the SOMEIP header). Depending on the system you are running on or the use case reserving the maximum amount of memory might be the right or the wrong choice. So I think this is a decision that the user of the library will have to make and the library should offer different options. Additionally the maximum size would get even bigger if someone chooses to use IPV6 Jumbograms, which would blow up the maximum size to ~4GiB, which would make allocating the maximum amount quiet impossible or at least costly. While it is unlikely that someone will combine SOMEIP with IPv6 Jumbograms, I learned that you should never assume anything and that at some point someone will do it.
> 
> TLDR: I am no longer against allocations, but I was thinking of flagging all parts of the library that need allocation or std types with the std feature flag (similar to https://github.com/BurntSushi/byteorder/blob/master/Cargo.toml#L25 ). That way users of the library can make a choice if they want to use those parts or explicitly exclude them.
> 
> So the plan would be that I review & merge your PR in the following weeks, maybe put the std feature flag on top of that and then do a release. Unless the std feature flagging takes me too long, then I will do a release just with your changes and the std feature on top later.
> 
> Greets Julian

Hi @JulianSchmid 

hope you feel better soon. There is no hurry for the review. Please take your time.

Seems like a reasonable approach to put this behind a feature flag. It would be great if there was a way to basically use e.g. const generics for systems that must avoid allocations and define reasonable constants (e.g. 65'535 byte) but also support the case where systems which are fine with dynamic memory allocations don't have to reserve 65'535 byte. The only way I see for that is feature flags. 

```Rust
const DEFAULT_MAX_ENTRIES: usize = ...;
const DEFAULT_MAX_OPTIONS: usize = ...;


#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct SomeIpSdHeader<const MAX_ENTRIES: usize, const MAX_OPTIONS: usize> {
    pub reboot: bool,
    pub unicast: bool,
    pub explicit_initial_data_control: bool,
    // reserved: [u8;3],
    // Length of entries array in bytes
    // length_of_entries: u32,
  
    #[cfg(not(feature = "std"))]
    pub entries: [SomeIpSdEntry; MAX_ENTRIES],
    #[cfg(feature = "std")]
    pub entries: Vec<SomeIpSdEntry>,
    
    // Length of entries array in bytes
    // length_of_options: u32,

    #[cfg(not(feature = "std"))]
    pub options: [SomeIpSdOption; MAX_OPTIONS],
    #[cfg(feature = "std")]
    pub options: Vec<SomeIpSdOption>,
}
```

The downside being that here for std, const generics would basically have to be defined although they are not used. I guess an alternative is completely replicating the struct. Anyway the downside is it obviously adds complexity, since one has to deal with multiple variants.

Feel free to take your time for the release like I said there is no hurry.

Best regards

Marcel

**Comment by JulianSchmid** [comment]
Hi, sorry for taking so long. I will add some minor changes on top of your PR and then release it as part of `0.5.0` today or tomorrow.
